lms/djangoapps/instructor_task/docs/decisions/001-scheduling-instructor-tasks.rst

```diff
@@ -0,0 +1,61 @@
+==========================
+Scheduled Instructor Tasks
+==========================
+
+Status
+------
+
+Draft
+
+Background
+----------
+
+One of the most requested features in edx-platform has been the ability to schedule an email message to be sent at a specified date and time through the Instructor Dashboard's Bulk Course Email tool. These email messages are sent as an instructor task. Today, edx-platform does not have the ability to support scheduling of instructor tasks. When an instructor task is created we attempt to execute the task immediately.
+
+This ADR covers details on the proposed modeling of a scheduled instructor task. A subsequent ADR will focus on the processing of scheduled instructor tasks.
+
+Decision
+--------
+
+* We will introduce a new ``InstructorTaskSchedule`` model to track the (optional) schedule of an instructor task.
+* We will extend the existing ``InstructorTask`` model with a new (nullable) column to link an instructor task to its schedule.
```

Comment:
Can we expand this to let the reader know if an InstructorTask is an item picked up by Celery worker or not? If it is such an item, what happens to the worker thread when the `cur_time` is < `due`?
If the task model is not going to kick off a celery worker item until after the due date, could we be specific on this?

Reply:
Be careful about timezones.  I assume you will store UTC, the workers will be UTC, but the display to the user needs to be localized.  (Sorry if that seems obvious.)

Reply:
Do you anticipate in the future having tasks that are relative to other date-timed things in the system?  Like courserun start/end or an assignment due date?  In those cases, would something know to move the `due` time on this model and persist it when the primary thing moves?  I'm not saying you have to include that at this stage, but I'm just wondering how you imagine it playing out.

Reply:
@hbwhbw Yep. This isn't reflected in the ADR but your assumptions match my assumptions. I'll try to make this clearer. The plan is to store as UTC. I hope for the backend data to receive the DateTime info in UTC from the Frontend/MFE.

Reply:
> Do you anticipate in the future having tasks that are relative to other date-timed things in the system? Like courserun start/end or an assignment due date? In those cases, would something know to move the `due` time on this model and persist it when the primary thing moves? I'm not saying you have to include that at this stage, but I'm just wondering how you imagine it playing out.

I'll give this some thought. I have been worrying about our immediate use case which is allowing course teams to schedule messages through the bulk course email tool.

Reply:
fwiw I believe Ryan has said this is a likely future use case. I don't think we need to design *for* it necessarily but maybe run through the thought experiment just to make sure that there wouldn't be anything obviously blocking it.